### PR TITLE
Don't consider a transaction expired until we've seen all collections it could have been in

### PR DESCRIPTION
**Pull request by jordanschalm**

### CODE REVIEW

File: engine/access/rpc/backend/backend_transactions.go [hunk 1]
```diff
@@ -245,15 +245,34 @@ func (b *backendTransactions) DeriveTransactionStatus(
...
+		// was included in a collection we just haven't seen yet.
+		//
+		// So - we check that a block has been finalized that expires the
+		// transaction (finalizedHeight-refHeight>EXPIRY) AND we check implicitly
+		// that no collection in the interim contains the transaction
+		// (fullHeight-refHeight>EXPIRY)
+		expiredWRTFinalized := finalizedHeight > refHeight && finalizedHeight-refHeight > flow.DefaultTransactionExpiry
+		expiredWRTFull := fullHeight > refHeight && fullHeight-refHeight > flow.DefaultTransactionExpiry
+
+		if expiredWRTFinalized && expiredWRTFull {
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_transactions.go:275]: should we nest the two conditions and then avoid the db lookup call `b.blocks.GetLastFullBlockHeight()` all together is `expiredWRTFinalized` is `false`?

File: engine/access/rpc/backend/backend_transactions.go [hunk 2]
```diff
@@ -245,15 +245,34 @@ func (b *backendTransactions) DeriveTransactionStatus(
...
 		finalized, err := b.state.Final().Head()
 		if err != nil {
 			return flow.TransactionStatusUnknown, err
 		}
+		finalizedHeight := finalized.Height
 
-		// Have to check if finalized height is greater than reference block height rather than rely on the subtraction, since
-		// heights are unsigned ints
-		if finalized.Height > referenceBlock.Height && finalized.Height-referenceBlock.Height > flow.DefaultTransactionExpiry {
+		// the last full height is the height where we have recieved all
```
Comment (vishalchangrani) [inline engine/access/rpc/backend/backend_transactions.go:256]: we can add a unit testcase for this - I can do it as a separate issue if you would like

**Review by vishalchangrani — APPROVED** [review]
lgtm
1. Thanks a lot for doing this!
2. Left you a couple of comments.

File: engine/access/rpc/backend/backend_transactions.go [hunk 3]
```diff
@@ -245,15 +245,34 @@ func (b *backendTransactions) DeriveTransactionStatus(
...
+		// collections for all blocks with a lower height
+		fullHeight, err := b.blocks.GetLastFullBlockHeight()
```
Comment (Kay-Zee) [inline engine/access/rpc/backend/backend_transactions.go:258]: We already had this function? Nice!

**Review by Kay-Zee — APPROVED** [review]

File: engine/access/rpc/backend/backend_transactions.go (hunk 1 above)
Comment (jordanschalm) [reply engine/access/rpc/backend/backend_transactions.go:275]: Added an early exit

File: engine/access/rpc/backend/backend_transactions.go (hunk 2 above)
Comment (jordanschalm) [reply engine/access/rpc/backend/backend_transactions.go]: done in https://github.com/onflow/flow-go/pull/390/commits/dc40f7e8af314577323d5d32607807937384a5d5

**Comment by jordanschalm** [comment]
Updated @Kay-Zee

File: engine/access/rpc/backend/backend_test.go [hunk 4]
```diff
@@ -387,20 +394,55 @@ func (suite *Suite) TestTransactionExpiredStatusTransition() {
...
+	suite.Run("expiry un-confirmed", func() {
 
-	// second call - reference block is now very far behind, and should be considered expired
-	result, err = backend.GetTransactionResult(ctx, txID)
-	suite.checkResponse(result, err)
+		suite.Run("ONLY finalized expiry block", func() {
+			// we have finalized an expiry block
+			headBlock.Header.Height = block.Header.Height + flow.DefaultTransactionExpiry + 1
+			// we have NOT observed all intermediary collections
+			fullHeight = headBlock.Header.Height - flow.DefaultTransactionExpiry/2
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply engine/access/rpc/backend/backend_test.go]: https://github.com/onflow/flow-go/pull/390/commits/8daf6de7660b2ea62563b1a527eddddd42c448c6

**Pull request merged**